File: src/ksp_cli/flat_velocity.py
"""The ``--flat-velocity`` grammar: ``fresh`` or a bare number, range-checked downstream."""

from ksp.midi_export import DEFAULT_FLAT_VELOCITY


def parse_flat_velocity(text: str | None) -> int | None:
    """Parse ``--flat-velocity``'s value into a velocity, unvalidated.
    A bare numeral passes through for :class:`ExportOptions` to range-check."""
    if text is None:
        return None
    if text == "fresh":
        return DEFAULT_FLAT_VELOCITY
    # Spelled out rather than left to ``int``, which also takes underscores and
    # non-ASCII digits: both cores must refuse exactly the same input.
    if not (text.isascii() and text.isdigit()):
        raise ValueError(f"--flat-velocity: '{text}' is not 'fresh' or a velocity")
    return int(text)

File: swift/Sources/KSPRun/FlatVelocity.swift
import KSPKit
import KSPMIDI

/// The word `--flat-velocity` takes for `MIDIExport.defaultFlatVelocity`.
public let freshVelocitySpec = "fresh"

/// Parse `text` into a velocity, unvalidated -- the options type does the range check.
public func parseFlatVelocity(_ text: String?) throws -> Int? {
    guard let text else { return nil }
    if text == freshVelocitySpec { return MIDIExport.defaultFlatVelocity }
    guard !text.isEmpty, text.allSatisfy({ $0.isASCII && $0.isNumber }) else {
        throw KSPError.value("--flat-velocity: '\(text)' is not 'fresh' or a velocity")
    }
    // A numeral too big for `Int` is still a number, so it earns the range message rather than
    // this one. Saturating to `.max` sends it there.
    return Int(text) ?? .max
}
